src/lib.rs

```diff
@@ -3605,6 +3620,115 @@ pub struct PartialResultParams {
     pub partial_result_token: Option<ProgressToken>,
 }
 
+#[derive(Debug, Eq, PartialEq, Default, Deserialize, Serialize, Clone)]
+#[serde(rename_all = "camelCase")]
+#[cfg(feature = "proposed")]
+pub struct SemanticHighlightingClientCapability {
+    /**
+     * `true` if the client supports semantic highlighting support text documents. Otherwise, `false`. It is `false` by default.
+     */
+    pub semantic_highlighting: bool,
+}
+
+#[derive(Debug, Eq, PartialEq, Default, Deserialize, Serialize, Clone)]
+#[cfg(feature = "proposed")]
+pub struct SemanticHighlightingServerCapability {
+    /**
+     * A "lookup table" of semantic highlighting [TextMate scopes](https://manual.macromates.com/en/language_grammars)
+     * supported by the language server. If not defined or empty, then the server does not support the semantic highlighting
+     * feature. Otherwise, clients should reuse this "lookup table" when receiving semantic highlighting notifications from
+     * the server.
+     */
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub scopes: Option<Vec<Vec<String>>>,
+}
+
+#[derive(Debug, Eq, PartialEq, Clone)]
+#[cfg(feature = "proposed")]
+pub struct SemanticHighlightingToken {
+    pub character: u32,
+    pub length: u16,
+    pub scope: u16,
+}
+
+#[cfg(feature = "proposed")]
+impl SemanticHighlightingToken {
+    /// Deserializes the tokens from a base64 encoded string
+    fn deserialize_tokens<'de, D>(deserializer: D) -> Result<Vec<SemanticHighlightingToken>, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let s = String::deserialize(deserializer)?;
+        let bytes =
+            base64::decode(s.as_str()).map_err(|_| serde::de::Error::custom("Error parsing base64 string"))?;
+        let mut res = Vec::new();
+        for chunk in bytes.chunks_exact(8) {
+            res.push(SemanticHighlightingToken {
+                character: (&chunk[0..4]).read_u32::<BigEndian>().unwrap(),
+                length: (&chunk[4..6]).read_u16::<BigEndian>().unwrap(),
+                scope: (&chunk[6..8]).read_u16::<BigEndian>().unwrap(),
+            });
+        }
+        Result::Ok(res)
+    }
+
+    /// Serialize the tokens to a base64 encoded string
+    fn serialize_tokens<S>(
+        tokens: &Vec<SemanticHighlightingToken>,
+        serializer: S,
+    ) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let mut bytes = vec![];
+        for token in tokens {
+            bytes.write_u32::<BigEndian>(token.character).unwrap();
```

Comment:
```suggestion
            bytes.extend_from_slice(&token.character.to_be_bytes());
```

etc. The byteorder crate should not be needed.